src/main/scala/workflow/Optimizer.scala

```diff
@@ -0,0 +1,35 @@
+package workflow
+
+/**
+ * Optimizes a Pipeline DAG
+ */
+object Optimizer extends RuleExecutor {
+  protected val batches: Seq[Batch] = Batch("DAG Optimization", FixedPoint(100), EquivalentNodeMerger) :: Nil
+}
+
+/**
+ * Merges equivalent node indices in the DAG.
+ * Nodes are considered equivalent if:
+ * - The nodes at the indices themselves are equal
+ * - They point to the same data dependencies
+ * - They point to the same fit dependencies
+ */
+object EquivalentNodeMerger extends Rule {
+  def apply[A, B](plan: Pipeline[A, B]): Pipeline[A, B] = {
+    val fullNodes = plan.nodes.zip(plan.dataDeps.zip(plan.fitDeps)).zipWithIndex
+    val mergableNodes = fullNodes.groupBy(_._1).mapValues(_.map(_._2)).toSeq
```

Comment:
Yeah I'll make that clearer in this line: ` - The nodes at the indices themselves are equal`